### DISCUSSION TITLE
Request for Inclusion of Lap Sector Times in API

### DISCUSSION THREAD

**Discussion by msfintech** [fabricepipart/mgp-timings#282] [Ideas]
Hello,
First, thank you for the amazing work. I have a request: could you add the lap sector times? These times are incredibly useful for understanding where a rider is gaining or losing against their competitors. You can find the sector times in this file: (https://resources.motogp.com/files/results/2024/GER/MotoGP/RAC/Analysis.pdf).

**Comment by fabricepipart** [comment]
Sure!
I think that should not be too hard as I already parse the analysis PDF. I'll give a look hopefully soon

**Comment by pmgfrederico** [comment]
That would be brilliant. This would come just in time when I am trying to do some sector time based analysis.

Looking at the API Spec looks like an opportunity to extended the _LapAnalysis_ schema with an optional SectorTimes section (without introducing any breaking changes), which is used in the following endpoint:
`/api/{year}/{eventShortName}/{category}/{shortSessionName}/analysis`

The fishiest thing is how badly behaved, structure wise, that particular pdf is but it really looks like that the code base already has a very effective parsing strategy in place!

**Comment by fabricepipart** [comment]
I hope I'll achieve this this week !

**Comment by fabricepipart** [comment]
Would it be useful ot have the SectorTimes in the lap analysis CSV or you dont care?

**Comment by msfintech** [comment]
Hi Fabrice, I recommend choosing the option that requires the least effort on your part.

**Reply by fabricepipart** [reply]
Then, I'll start with JSON part. It is mostly done in PR #288. I'll add tests and polish this week :)

**Comment by fabricepipart** [comment]
Here you are !
You now have the sectors in the analysis output.

For example Assen Moto GP Fp2 this year:
https://mgp-timings.evo.teknichrono.fr/api/2024/NED/MOTOGP/FP2/analysis

**Reply by pmgfrederico** [reply]
Not looking to be picky, but from a completeness perspective, would you consider supporting which sector time was cancelled, which ultimately will get the lap cancelled. Is it an asterisk that they use also close to the sector time?

**Reply by msfintech** [reply]
Hi Fabrice,

All the spot checks I conducted indicate that the data is correct, so the initial request has been fulfilled. 

If possible and if your time allows, I noticed that the PDF contains additional useful information that could enrich the dataset, such as:
Total Runs, Total Laps, Full Laps, Valid Laps (12), Run Number, and, as mentioned by Pedro, the extraction of flags for the cancelled sector (*). 

It would also be perfect if the output could be available as a CSV (although this is not very important). 

Would you like me to log this as a new request?

**Reply by fabricepipart** [reply]
This text is enough as a request
I think I already fill the information if the lap is cancelled

I am unsure I'll be able to do it this week though.
Probably during the August month (I am off for 2w after this week)

**Reply by fabricepipart** [reply]
After looking better at it, I think I'll change the data structure.
I'll introduce (to avoid repeating data) a RiderAnalysis that contains several Runs that contains several Laps that contains several Sectors.
What do you think ?

**Reply by msfintech** [reply]
Hi Fabrice, having all the information in one place seems like a good idea. No need to rush; take your time. I have some more ideas that I'll share in a separate comment to keep things organized.

**Reply by pmgfrederico** [reply]
The former structure is probably more oriented towards a tabular view or for CSV content negotiation.

Something in the lines of this?

```
components:
  
  schemas:
      
    SectorTime:
      type: object
      properties:
        number:
          type: integer
          minimum: 1
          maximum: 4
        durationInMs:
          type: integer
        isCancelled:
          type: boolean
      required:
        - number
        - durationInMs
        - isCancelled
    
    LapTime:
      type: object
      properties:
        number:
          type: integer
        durationInMs: # could be ommited via the projection of the sum of the sector times
          type: integer
        sectorTimes:
          type: array
          items:
            $ref: '#/components/schemas/SectorTime'
          minItems: 0
          maxItems: 4
        topSpeedInKph:
          type: number
          format: float
      required:
        - number
        - durationInMs
        - sectorTimes
        - topSpeedInKph
        
    Tyre:
      type: object
      properties:
        compound:
          type: string #could be an enum instead...
        startingLaps:
          type: string
          minimum: 0 # 0 = New Tyre
      required:
        - compound
        - startingLaps
    
    Run:
      type: object
      properties:
        number:
          type: integer
        laps:
          type: array
          items:
            $ref: '#/components/schemas/LapTime'
        frontTyre:
          $ref: '#/components/schemas/Tyre'
        rearTyre:
          $ref: '#/components/schemas/Tyre'
      required:
        - number
        - laps
        - frontTyre
        - rearTyre
        
    RiderSession:
      type: object
      properties:
        rider:
          $ref: '#/components/schemas/Rider'
        classification:
          type: integer
          minimum: 1
        runs:
          type: array
          items:
            $ref: '#/components/schemas/Run'
          minItems: 0
      required:
        - rider
        - classification
        - runs
          
    RiderSessionExtended:
      allOf:
        - $ref: '#/components/schemas/RiderSession'
        - type: object
          properties:
            totalLaps:
              type: integer
            fullLaps:
              type: integer
            validLaps:
              type: integer
          required:
            - totalLaps,
            - fullLaps,
            - validLaps
          
    Rider: # just a placeholder
      type: object
```

**Reply by fabricepipart** [reply]
Yes I think that's exactly what I plan to do
For the sake of backward compatibility I'll keep both (anyway the other structure is adapted for the CSVs as mentioned).

I'll probably name the new entrypoint `/sectors`

**Reply by pmgfrederico** [reply]
Resource wise, its still the same - _sessionAnalysis_. By fully introducing a new response model you will be introducing a breaking change.

I see two options to keep the same resource name:

- Introducing a version segment to your endpoints: /api/**v2**/{year}/{eventShortName}/{category}/{shortSessionName}/analysis an support both endpoints side by side
- or, if the current endpoint only support a text/csv mime type, you could use content negotiation and define a different model for application/json mime type

**Comment by msfintech** [comment]
Fantastic job, Fabrice! I need to do some work with the JSON on my end to test it, but I'll let you know how it goes.

**Comment by msfintech** [comment]
As I mentioned, I was looking through the available PDFs and found three more that caught my attention. They contain interesting information worth analyzing.
https://resources.motogp.com/files/results/2024/GER/MotoGP/RAC/LapChart.pdf
https://resources.motogp.com/files/results/2024/GER/MotoGP/EntryBiographical.pdf
https://resources.motogp.com/files/results/2024/GER/MotoGP/table1.pdf
Again, you've done a fantastic job. There's no need to rush with any of these requests; they are just nice to have.

**Reply by fabricepipart** [reply]
WOuld you mind opening a dedicated issue for that? It is just for me to get organized :)

**Reply by msfintech** [reply]
I'll do Fabrice
